code/game/objects/items/robot/robot_upgrades.dm

```diff
@@ -343,93 +402,79 @@
-	. = ..()
-	if (.)
-		if(toggle_action)
-			QDEL_NULL(toggle_action)
-		cyborg = null
-		deactivate_sr()
-
-/obj/item/borg/upgrade/selfrepair/dropped()
-	..()
-	addtimer(CALLBACK(src, .proc/check_dropped), 1)
-
-/obj/item/borg/upgrade/selfrepair/proc/check_dropped()
-	if(loc != cyborg)
-		if(toggle_action)
-			QDEL_NULL(toggle_action)
-		cyborg = null
-		deactivate_sr()
+		action = new /datum/action/item_action(src)
+		action.name = name
+		action.Grant(R)
 
 /obj/item/borg/upgrade/selfrepair/ui_action_click()
-	on = !on
-	if(on)
+	if(COOLDOWN_FINISHED(src, recharging_repmod) && icon_state == "selfrepair_off")
+		icon_state = "selfrepair_on"
 		to_chat(cyborg, "<span class='notice'>You activate the self-repair module.</span>")
-		START_PROCESSING(SSobj, src)
+		playsound(cyborg.loc, activation_sound, 30)
+		update_icon()
+		process()
 	else
-		to_chat(cyborg, "<span class='notice'>You deactivate the self-repair module.</span>")
-		STOP_PROCESSING(SSobj, src)
-	update_icon()
+		to_chat(cyborg, "<span class='notice'>Your self-repair module is not ready to be activated again yet.</span>")
 
 /obj/item/borg/upgrade/selfrepair/update_icon()
+	. = ..()
 	if(cyborg)
-		icon_state = "selfrepair_[on ? "on" : "off"]"
-		for(var/X in actions)
-			var/datum/action/A = X
-			A.UpdateButtonIcon()
+		if(timer_overlay_active && COOLDOWN_FINISHED(src, recharging_repmod))
+			end_timer_animation()
+		if(action)
+			action.UpdateButtonIcon()
 	else
 		icon_state = "cyborg_upgrade5"
 
+
+
 /obj/item/borg/upgrade/selfrepair/proc/deactivate_sr()
 	STOP_PROCESSING(SSobj, src)
-	on = FALSE
-	update_icon()
-
-/obj/item/borg/upgrade/selfrepair/process()
-	if(world.time < next_repair)
-		return
 
-	if(cyborg && (cyborg.stat != DEAD) && on)
-		if(!cyborg.cell)
-			to_chat(cyborg, "<span class='warning'>Self-repair module deactivated. Please, insert the power cell.</span>")
-			deactivate_sr()
-			return
-
-		if(cyborg.cell.charge < powercost * 2)
-			to_chat(cyborg, "<span class='warning'>Self-repair module deactivated. Please recharge.</span>")
-			deactivate_sr()
-			return
-
-		if(cyborg.health < cyborg.maxHealth)
-			if(cyborg.health < 0)
-				repair_amount = -2.5
-				powercost = 30
+/obj/item/borg/upgrade/selfrepair/proc/cooldown_start(ticksleft = 0)
+	icon_state = "selfrepair_off"
+	update_icon()
+	playsound(cyborg.loc, deactivation_sound, 60)
+	COOLDOWN_START(src, recharging_repmod, ((20 SECONDS) + ((100 SECONDS) - ((10 SECONDS) * ticksleft)))) //20 seconds + 10 seconds for every repair tick used up. 2 minutes max
+	begin_timer_animation()
+
+/obj/item/borg/upgrade/selfrepair/process(delta_time)
+	if(!cyborg) //Sanity check to reset the module in case it is somehow removed while running. 
+		update_icon()
+		COOLDOWN_RESET(src, recharging_repmod)
+		repair_ticks = 10
+		deactivate_sr()
+	if(COOLDOWN_FINISHED(src, recharging_repmod) && icon_state == "selfrepair_on")
+		while(repair_ticks)
+			sleep(10)
```

Comment:
Probably better to use a timer here, sleeps cause a lot of jank because they cause anything calling this function to also be delayed, which can lead to problems.

Reply:
Awaiting response on what exactly a timer is from coding channel. Will do once I know what they are.

Reply:
I've moved the sleep proc outside of process because it didn't need to be there anyway.